src/ngraph/node.hpp

```diff
@@ -324,6 +376,309 @@ namespace ngraph
+        void replace_source_output(const Output<Node>& new_source_output) const;
+
+        bool operator==(const Input& other) const
+        {
+            return m_node == other.m_node && m_index == other.m_index;
+        }
+        bool operator!=(const Input& other) const
+        {
+            return m_node != other.m_node || m_index != other.m_index;
+        }
+        bool operator<(const Input& other) const
+        {
+            return m_node < other.m_node || (m_node == other.m_node && m_index < other.m_index);
+        }
+        bool operator>(const Input& other) const
+        {
+            return m_node > other.m_node || (m_node == other.m_node && m_index > other.m_index);
+        }
+        bool operator<=(const Input& other) const
+        {
+            return m_node <= other.m_node || (m_node == other.m_node && m_index <= other.m_index);
+        }
+        bool operator>=(const Input& other) const
+        {
+            return m_node >= other.m_node || (m_node == other.m_node && m_index >= other.m_index);
+        }
+
+    private:
+        NodeType* const m_node;
+        const size_t m_index;
+    };
+
+    /// \brief A handle for one of a node's outputs.
+    template <typename NodeType>
+    class Output
+    {
+    public:
+        /// \brief Constructs a Output.
+        /// \param node A pointer to the node for the output handle.
+        /// \param index The index of the output.
+        Output(NodeType* node, size_t index)
+            : m_node(node)
+            , m_index(index)
+        {
+        }
+
+        /// \brief Constructs a Output.
+        /// \param node A `shared_ptr` to the node for the output handle.
+        /// \param index The index of the output.
+        ///
+        /// TODO: Make a plan to deprecate this.
+        Output(const std::shared_ptr<NodeType>& node, size_t index)
+            : m_node(node.get())
+            , m_index(index)
+        {
+        }
+
+        /// \brief Constructs a Output, referencing the zeroth output of the node.
+        /// \param node A `shared_ptr` to the node for the output handle.
+        template <typename T>
+        Output(const std::shared_ptr<T>& node)
+            : Output(node, 0)
+        {
+        }
+
+        /// \return A pointer to the node referred to by this output handle.
+        NodeType* get_node() const { return m_node; }
+        /// \return A `shared_ptr` to the node referred to by this output handle.
+        ///
+        /// TODO: Make a plan to deprecate this.
+        std::shared_ptr<NodeType> get_node_shared_ptr() const { return m_node->shared_from_this(); }
+        /// \return The index of the output referred to by this output handle.
+        size_t get_index() const { return m_index; }
+        /// \return A reference to the tensor descriptor for this output.
+        descriptor::Tensor& get_tensor() const
+        {
+            return m_node->m_outputs.at(m_index).get_tensor();
```

Comment:
Use node methods to get access to output. Same for all other methods in this class which directly access Node's innards

Reply:
(Same comment)